src/main.rs

```diff
@@ -651,6 +691,219 @@ fn should_initialize_git_repo(mut path: &Path) -> bool {
+            log::warn!(
+                "`rustc +{name} --print target-list` exited with status {:?}",
+                output.status.code()
+            );
+            continue;
+        }
+
+        if !String::from_utf8_lossy(&output.stdout)
+            .lines()
+            .any(|l| l.trim() == target)
+        {
+            // target not found - skip
+            continue;
+        }
+
+        // call `rustc +<toolchain> --version` and compare to `msrv`
+        if let Some(ver) = check::get_version("rustc", &[&format!("+{name}")]) {
+            if !ver.is_at_least(msrv) {
+                // toolchain version is below MSRV - skip
+                continue;
+            }
+        } else {
+            log::warn!(
+                "Failed to detect rustc version for toolchain `{name}`; skipping MSRV check"
+            );
+        }
+
+        available.push(name.to_string());
+    }
+
+    Ok(available)
+}
+
+/// This functiomn will find the `toolchain` category in `template.yaml` and replace its placeholder option with one
+/// option per installed rustup toolchain that supports required target.
+fn populate_toolchain_category(
+    chip: Chip,
+    options: &mut [GeneratorOptionItem],
+    cli_toolchain: Option<&str>,
+    msrv: &check::Version,
+) -> Result<()> {
+    let target = chip.target().to_string();
+
+    let mut available = filter_toolchains_for(&target, msrv)?;
+
+    // for now, we should hide the generic toolchains for Xtensa (stable-*, beta-*, nightly-*).
+    if chip.is_xtensa() {
+        available.retain(|name| {
+            !(name.starts_with("stable") || name.starts_with("beta") || name.starts_with("nightly"))
+        });
+    }
+
+    // sanity check
+    if available.is_empty() {
+        if let Some(cli) = cli_toolchain {
+            if chip.is_xtensa()
+                && (cli.starts_with("stable")
+                    || cli.starts_with("beta")
+                    || cli.starts_with("nightly"))
+            {
+                bail!(
+                    "Toolchain `{cli}` is not supported for Xtensa targets; \
+                     please use different toolchain (e.g. `esp`)"
+                );
+            }
+
+            bail!(
+                "Toolchain `{cli}` does not have target `{target}` installed (or no toolchain does)"
+            );
+        }
+        log::warn!(
+            "No rustc toolchains found that have `{target}` installed; toolchain category will stay as placeholder"
+        );
+        return Ok(());
+    }
+
+    if let Some(cli) = cli_toolchain {
+        if !available.iter().any(|t| t == cli) {
+            if chip.is_xtensa()
+                && (cli.starts_with("stable")
+                    || cli.starts_with("beta")
+                    || cli.starts_with("nightly"))
+            {
+                bail!(
+                    "Toolchain `{cli}` is not supported for Xtensa targets; \
+                     please use an ESP toolchain (e.g. `esp`)"
```

Comment:
Maybe we should link to the book https://docs.espressif.com/projects/rust/book/getting-started/toolchain.html#xtensa-devices